### DISCUSSION TITLE
2024-07-05 - Architectural Decision Records (ADRs) - Chewing the Fat 📈

### DISCUSSION THREAD

**Discussion by bradystroud** [SSWConsulting/SSW.Rules.Content#8834] [Chewing the Fat]
https://www.ssw.com.au/rules/architectural-decision-records/

Are you a fan of ADRs?
✅ What kind of decisions would you record?
❌ What kind of decisions would you **not** record?

**Comment by JeanThirion** [comment]
> Are you a fan of ADRs?

YES - absolutely love it

> ✅ What kind of decisions would you record?

Everything architecture related - I like how Dan puts it "everything that will be expensive to revert"

> ❌ What kind of decisions would you not record?

I guess all the others; things decided on the fly, or things discussed with the client (i.e. business decisions)


--- 

This gave me an idea: why don't we build a tool like Log4Brains but for Business decisions?
Inputs would be meeting transcripts and various notes, put through GPT magic, and output would be some markdown. Add a static site to render something like L4B... and voila; you've got a track record of business decisions. 

That would have literally saved me countless hours with Bupa or other.. ehm... "dynamic and challenging" client environments.

**Reply by drwharris** [reply]
> This gave me an idea: why don't we build a tool like Log4Brains but for Business decisions? Inputs would be meeting transcripts and various notes, put through GPT magic, and output would be some markdown. Add a static site to render something like L4B... and voila; you've got a track record of business decisions.
> 
> That would have literally saved me countless hours with Bupa or other.. ehm... "dynamic and challenging" client environments.

Why couldnt you use Log4Brains?

**Reply by matt-goldman-ssw** [reply]
> Why couldnt you use Log4Brains?

It's pretty dev focused. Would need to Tinafy it.

**Reply by danielmackay** [reply]
Had a brief discussion with @bradystroud and @adamcogan about using Tina for ADRs.  Although it's possible, it won't be as straight forward to setup and use as L4B is.

**Comment by PothieuG** [comment]
Very fan of it !

I would record decisions about:
- architecture
- database
- security
- tech stack
- devops/deployment strategy

I wouldn't record decisions about :
- best practice
- testing strategy

**Comment by Freego1783** [comment]
> Are you a fan of ADRs?

Was not aware this was a thing, but yeah definitely!

> ✅ What kind of decisions would you record?

Everything related to the architecture or the tech stack

> ❌ What kind of decisions would you not record?

Everything else

**Comment by ricksu978** [comment]
> Are you a fan of ADRs?

Yes

> ✅ What kind of decisions would you record?
> ❌ What kind of decisions would you not record?

I would definitely record "unusual" decisions.
e.g.
I wouldn't record why we use C# in a project, because that what we all use in SSW.
But if we decided to use VB.Net or F#, I'll definitely record that.

**Comment by Levijj22** [comment]
Great Topic!
The video was super easy to follow.

I like that we have standard terminology for this kind of thing.
Thinking about how we can use this concept in the sales world.

**Comment by tiagov8** [comment]
Big fan of good communication.

The challenge is to find the **right balance** of defining what is important to have ADRs. Documenting every decision might be overwhelming and counter-productive. I guess each project will have different ways of dealing with this.

**Comment by drwharris** [comment]
❓Are you a fan of ADRs?
Yes, love them
 
✅ What kind of decisions would you record?
Anything that 
- would consistently be inconsistent, 
- causes repeated discussion, 
- is contentious, 
- has significant financial impact

❌ What kind of decisions would you not record?
Things that aren't Architecture related, eg code styles

**Reply by GordonBeeming** [reply]
So you're saying 2 vs 4 spaces wouldn't be in the ADRs 😅

**Comment by tkapa** [comment]
ADRs seem great. They make a lot of sense, and can help with onboarding and recalling infromation.

As others have said above, pretty much anything that seems like a "point of no return". If you have to have a meeting about it, or a decision needs to be made by a stakeholder, that sounds like something that could be logged in an ADR.

**Comment by matt-goldman-ssw** [comment]
> Are you a fan of ADRs?

Absolutely

> ✅ What kind of decisions would you record?

Primarily architecture decisions, but fundamentally anything that would be useful for people to refer back to and understand why a decision was made, _particularly_  when selecting one option over others.

> ❌ What kind of decisions would you not record?

Really anything that doesn't meet the above criteria. For example I wouldn't document why I cast something to a list and then iterated over it rather than querying each entity separately. (Hugely contrived example).

**Comment by ozdemir-mehmet** [comment]
Are you a fan of ADRs?
✅ What kind of decisions would you record?

What others lie Andrew H or Tylah have said.

❌ What kind of decisions would you not record?

Everything else

**Comment by leahy268** [comment]
> Are you a fan of ADRs?
Yes, it's important to know why you made a decision

> ✅ What kind of decisions would you record?
Anything important.  Not sure I'm qualified enough to speak to this one.

> ❌ What kind of decisions would you not record?
Not really sure.  But things that can easily be changed later I guess.

**Comment by Tanya-Leahy** [comment]
Are you a fan of ADRs?
Definitely! Great way of recording what and why decisions were made. 

✅ What kind of decisions would you record?
Any decisions that were made that impacted the project in a big way. Great to have a historical record!

❌ What kind of decisions would you not record?
Not sure as I haven't used ADRs before. Maybe decisions that could be easily reversed.

**Comment by jakebayliss** [comment]
Are you a fan of ADRs?
Yes

✅ What kind of decisions would you record?
Technology and archecture decisions mostly. Anything that is hard to reverse 

❌ What kind of decisions would you not record?
Low-impact decisions

**Comment by calumjs** [comment]
ADRs are cool. One thing I don't like is duplicated effort of creating the file - perhaps YakShaver should be able to generate them @uly1?
Alternatively we could have a GPT monitoring PBIs (and emails) and auto-generating suggested ADR drafts - a job for EagleEye perhaps @piers-sinclair?
We could also add a chatbot to the log4brains static site with access to the ADRs, code, PBIs, etc., which becomes the project assistant.

**Reply by danielmackay** [reply]
Love the idea about the L4B chatbot!

**Comment by joshbermanssw** [comment]
> Are you a fan of ADRs?

Yes 

> ✅ What kind of decisions would you record?

I'm +1'ing all the people that said this before me - but i think its vital to have a clear line of what decisions should be documented, in the documentation. Like Dan put it "everything that will be expensive to revert" 

> ❌ What kind of decisions would you not record?

Insignificant decisions or decisions that are self-explanatory

**Comment by ncn-ssw** [comment]
> Are you a fan of ADRs?

❌ No - it is a separate set of documentation to check. I only see it being useful when work is not being properly documented in PBIs.

Whenever I've wondered why a decision was made, I've used `git blame` to commence tracing the PR to the PBI. In the codebases I've been involved with at SSW, there's normally a Spike PBI to allocate development resources to investigating issues and making a decision. I have found it more convenient to have the decision documented on those PBIs.

> ✅ What kind of decisions would you record?

- Selection of service dependencies
- Addition/removal of libraries
- High level sub-system design

> ❌ What kind of decisions would you not record?

- Selection of individual variable and method names
- Selection of non-critical algorithms

**Reply by danielmackay** [reply]
@ncn-ssw . Why not do both - Create an ADR and link it into the PBI?

For long running projects when you have thousands of PBIs, I think it can be difficult to locate the PBI that contains the decision.

The thing with ADRs is that it's a single place to look for all important decisions, making the information much easier to find.

**Comment by edkranz** [comment]
> Are you a fan of ADRs?

Yes. It's a good place to document decisions which are not immediately clear from reading the codebase.

> ✅ What kind of decisions would you record?

Any decision which would deviate from normal patterns, with no apparent reason in code — normally made due to specific implementation requirements (significant enough to affect implementations of other things)

> ❌ What kind of decisions would you not record?

Decisions which are either trivial, or ones that adhere to existing software design patterns.

**Comment by chrisschultzssw** [comment]
> Are you a fan of ADRs?

I am now! 

> ✅ What kind of decisions would you record?

"ADRs should be used for anything **Important**" - _Daniel Mackay_

> ❌ What kind of decisions would you not record?

Anything unimportant or easy to reverse.

**Comment by stephanfakossw** [comment]
N/a - not a dev

**Comment by tomek-i** [comment]
>Are you a fan of ADRs?

Yes


>✅ What kind of decisions would you record?

Everything that is important, expensive to change, or if there have been different suggestions moving forward between devs
or something that deviates from the norm/best practice. 

>❌ What kind of decisions would you not record?

Everything else that's not worth recording.

**Comment by awaltos** [comment]
Any structured tool that helps document a project better is likely a good thing. This system seems like its low effort and easy to maintain 👍 I would lean to using docs that are already in the project as long as they are suitable to record ADR. E.g. if there is already a Confluence, use that instead of bringing in another tool like L4B.

> ✅ What kind of decisions would you record?

High level things. Any time a flow diagram is affected by a decision. Change of vendor. Etc

> ❌ What kind of decisions would you not record?

As others have said, trivial things. Things that don't impact other areas of the project

**Comment by steven0x51** [comment]
✅ What kind of decisions would you record?
High level decisions such as the architecure
❌ What kind of decisions would you not record?
The things that many change on the fly, such as class name, API name

**Comment by Ben0189** [comment]
Are you a fan of ADRs?
- Yes

✅ What kind of decisions would you record?
- Decision that is expensive to revert

❌ What kind of decisions would you not record?
- low impact decision (e.g. small package/lib used in the project)
- temporary decision (e.g. workaround for some uses case)

**Comment by bettybondoc** [comment]
I'm a fan of proper documentation, so ADRs are a plus in my book! I've seen them being drafted and used firsthand in a current client project, and I can definitely see the benefits.

**Comment by Calinator444** [comment]
> Are you a fan of ADRs?

I didn't know about them, I need to use them now

> ✅ What kind of decisions would you record?

- Architecture decisions
- Choice of external dependencies (whether information is being retrieved from SharePoint, a custom endpoint etc.)

> ❌ What kind of decisions would you not record?

There's a seemingly endless list, but really anything that won't have a major impact throughout the projects' lifecycle, e.g.:
- Choosing a variable name
- Choosing a naming scheme for endpoint urls
- whether a specific method in the source code uses recursion over iteration

**Comment by christoment** [comment]
> Are you a fan of ADRs?

Yes, documenting important decisions helps a lot in the future!

> ✅ What kind of decisions would you record?

Infrastructure changes, architecture changes

> ❌ What kind of decisions would you not record?

- Trivial technical implementation
- Things that will change frequently
- Business logic changes (this should be in a PBI, recorded and signed-off by the product owner!)

**Comment by MattParkerDev** [comment]
> Are you a fan of ADRs?

Yes - super important for recording decisions that could be questioned in the future.
> ✅ What kind of decisions would you record?

Large decisions, important decisions, contentious decisions.
> ❌ What kind of decisions would you not record?

Like Dan's video said - things like names of classes etc

**Comment by cvionnet** [comment]
As people move from one project to another, ADRs help prevent the loss of important information and decisions. They are the memory of the project 🐘. One of the challenges is to keep them up to date as the system evolves.

The Decision Log greatly enhances readability and provides a quick overview.

> ✅ What kind of decisions would you record?

Architectural choices, infrastructure decisions, specific libraries used, and any information intended for long-term use.

> ❌ What kind of decisions would you not record?

Trivial or temporary decisions and implementation

**Comment by miichaelsmedley** [comment]
> Are you a fan of ADRs? 
Super good idea - yes a fan.

✅ What kind of decisions would you record? 
Everything that led to all the major, decisions on a project

❌ What kind of decisions would you **not** record?
details or operational stuff.

**Comment by AttackOnMorty** [comment]
> Are you a fan of ADRs?

Yes - Good to have important decisions recorded in a centralized place.
 
Devs sometimes even forget the code they wrote 3 days ago, let alone these decisions.

**Comment by bahjatalaadelssw** [comment]
> Are you a fan of ADRs?
Yes, any kind of paper trail is important to accountability

> ✅ What kind of decisions would you record?
One's that would change to logic of a project. In video it would be like music or narrative

> ❌ What kind of decisions would you not record?
Minor elements, e.g. text, colour correction or frame composition

**Comment by Marxoz** [comment]
Are you a fan of ADRs?
✅ What kind of decisions would you record?
- Architectural decisions
- Decisions that can affect security and privacy
- Decisions that have $ associated with it

❌ What kind of decisions would you not record?
- Anything that can be passed as a requirement in a PBI

**Comment by will-truscott-ssw** [comment]
Are you a fan of ADRs?
> Yep! Noting down decisions not only ensures important information carries over to other team members, it also keeps devs accountable for major decisions made (especially esoteric ones) in a way that allows them to provide reasoning.

✅ What kind of decisions would you record?
> Anything that may slow or derail a project's progress if not known, or those you feel future devs may be curious about or confused over. Maintaining developer confidence in a system is incredibly important as it helps devs maintain a critical eye rather than getting in a mindset where they start dismissing issues because they view the whole codebase as subpar.

❌ What kind of decisions would you not record?
> I believe documentation should stay slim and targeted. I feel any decisions that speak for themselves or are fairly standard may not benefit the team to be made into an ADR, keep it to the important, the interesting or the controversial so devs actually read the docs.

**Comment by AntPolkanov** [comment]
> Are you a fan of ADRs?

Yes, they are also great for onboarding new developers as it gives an overview of the technologies used.

> ✅ What kind of decisions would you record?

Architectural decisions, library / framework decisions.

> ❌ What kind of decisions would you not record?

Naming conventions.

**Comment by jeoffreyfischer** [comment]
This is my first time learning about ADRs. 
I'm a firm believer of documenting - so I'll definitely use ADRs in the future 👍

**Comment by SamBWagner** [comment]
> Are you a fan of ADRs?

I like the idea, I just do worry about adding MORE documentation. I have no problem with documentation, it's just that there is an infinite well of reasons to add it and there's seldom good consideration for the upkeep and cost of doing it.

I think ADRs are pretty awesome. However, what are the reasons to have them over something like Architecture Documents? I can list a fair few like: Cost of maintenance, Cost of creation and Time to becoming outdated. However, ADRs aren't immune to this criticism either. I think some of the reasons to use them are:

1. They're cheap to create
2. They're cheap to maintain (if not abused)
3. They're cheap to throw out (if stuff changes)
4. They're cheap to parse

Even with these points though, it's always a cost/benefit thing. Not every project needs these kinds of records. Some might need something more verbose. It depends.

I think ADR's are a great tool to add to the toolbelt. Not a silver bullet though.

> ✅ What kind of decisions would you record?

I'd document anything that will have direct cost to the client and/or future developers. I would prioritise being polite to both your client and the people who come after you. If this is done via an ADR or an architecture document or a code comment is irrelevant to me. So long as whatever form of documentation you use is the most polite it can be.

> ❌ What kind of decisions would you not record?

Anything that doesn't fall squarely into the above bucket. If it doesn't cause meaningful harm (money loss) to the client, then don't document it. Just build it. If the code you're writing, tool you're using, library you've chosen wont cause the developers who are following you any meaningful harm then you need not document that either.

**Comment by JackDevAU** [comment]
✅ I like them a bunch! but I also think this can create a lot of overhead. I would follow the following when deciding if we should create an ADR

Easily Reversible - ❌ No ADR
Not Easily Reversible - ✅ Yes ADR 

![image](https://github.com/SSWConsulting/SSW.Rules.Content/assets/57518417/e4b24cba-132e-4a3c-a60b-77325e77bd52)
**Figure: Jeffs decision making**

**Reply by danielmackay** [reply]
Excellent insight @JackDevAU!

**Comment by danielmackay** [comment]
> Are you a fan of ADRs?

💯 

> ✅ What kind of decisions would you record?

Important decisions, that have long lasting affects.  Or any decision where I spent time discussing options with colleagues.

> ❌ What kind of decisions would you not record?

'Two-way door' decisions (easily reversible) or decisions where there is a clear choice

**Comment by MarcusIrmscherSSW** [comment]
As a non-dev I am just learning about ADR's for the first time. It seems like these would be an invaluable tool/method.

**Comment by JerryLuo5799** [comment]
> Are you a fan of ADRs?

Yes.

> ✅ What kind of decisions would you record?

Decisions that have a significant impact on the project, such as options of Solution, architecture, solution changes and so on.

> ❌ What kind of decisions would you not record?

Something the team defaults to,  etc. software development standards.

**Comment by KristenHu** [comment]
> Are you a fan of ADRs?

Yes, it looks great
> ✅ What kind of decisions would you record?

Any unusual decision related to the project. 
> ❌ What kind of decisions would you not record?

I wouldn't record the thing if it has a conventional standard or a clear decision.

**Comment by jimmidier** [comment]
> Are you a fan of ADRs?

Yes I am
  > ✅ What kind of decisions would you record?

I would record general decisions that would impact product roadmaps or work flows
  > ❌ What kind of decisions would you not record?

I would not record tech details decisions that could be in the documentations instead

**Comment by Aibono1225** [comment]
> Are you a fan of ADRs?

Yes

> ✅ What kind of decisions would you record?

Major decisions, architecture, security, etc..

❌ What kind of decisions would you not record?

Reversible decisions

**Comment by babakamyljanovssw** [comment]
> Are you a fan of ADRs?

Yes

> ✅ What kind of decisions would you record?

Architecture
Cloud Infrastructure
Tech Stack

> ❌ What kind of decisions would you not record?

Coding/Tech details

**Comment by PennyWalker** [comment]
I think anything important or unchangeable should be documented, but kept it lean and useful :)

**Comment by amankumarrr** [comment]
>Are you a fan of ADRs?

Yes

> ✅ What kind of decisions would you record?

Major decisions, locked npm libraries, major impacting areas

**Comment by ozair-ashfaque** [comment]
> ✅ What kind of decisions would you record?

Architecture decisions

> ❌ What kind of decisions would you not record?

New Features or Refactoring

**Comment by YalalCakirSSW** [comment]
> Are you a fan of ADRs?

I am now. 

> ✅ What kind of decisions would you record?

Important Features, Architecture

> ❌ What kind of decisions would you not record?

Small features

**Comment by isaaclombardssw** [comment]
> Are you a fan of ADRs?

I am not familiar with them, but the concept is great.

> ✅ What kind of decisions would you record?

Choices between architectures, libraries (if non-obvious), services being used.

> ❌ What kind of decisions would you not record?

Implementation, test strategy or reporting, other technical details.

**Comment by Dhruv-0987** [comment]
Are you a fan of ADRs?
Yes they are very important, specially in client projects
✅ What kind of decisions would you record?
Most of the time what 3rd party libraries we are using to do what and sometimes a design pattern which would be beneficial to the project  
❌ What kind of decisions would you not record?
Implementation details

**Comment by tino-liu** [comment]
> ✅ What kind of decisions would you record?

Big decisions like architecture and anything that makes or breaks the application

> ❌ What kind of decisions would you not record?

Small features

**Comment by TobyChurches** [comment]
> Are you a fan of ADRs?
 
 Yes! They are great - particularly for new Devs on a project
 
> ✅ What kind of decisions would you record?

Important ones 😂 All decisions that shape the direction of the project or that could need justification later

> ❌ What kind of decisions would you not record?

Implementation Details

**Comment by vladislav-kir** [comment]
> Are you a fan of ADRs?

Yes - definitely!

> ✅ What kind of decisions would you record?

Architecture, Tech stack, DevOps.

> ❌ What kind of decisions would you not record?

Implementation, Testing strategies.
